### ISSUE TITLE
Elastic BeanStalk deployments

### ISSUE THREAD

**Issue by FeodorFitsner** [appveyor/ci#45]
Discussion: http://help.appveyor.com/discussions/questions/409-deploying-to-amazon-elastic-beanstalk

Help:
http://docs.aws.amazon.com/AWSToolkitVS/latest/UserGuide/tkv-deployment-tool.html
http://docs.aws.amazon.com/elasticbeanstalk/latest/dg/deploy_NET_standalone_tool.html

**Comment by tbenade** [comment]
+1

**Comment by mwhouser** [comment]
+1

**Comment by danielmcormond** [comment]
Yes as @FeodorFitsner mentioned above, you can already deploy to Elastic Beanstalk using the [awsdeploy.exe](http://docs.aws.amazon.com/elasticbeanstalk/latest/dg/deploy_NET_standalone_tool.html) deployment tool. In your _appveyor.yml_ `deploy_script:` just do:

```
C:\Program Files (x86)\AWS Tools\Deployment Tool\awsdeploy.exe /redeploy /wait /DAWSAccessKey=%AWSAccessKey% /DAWSSecretKey=%AWSSecretKey% path\to\your\awsdeploy-options.txt
```

I'd suggest encrypting your AWS credentials with AppVeyor [encrypt tool](https://ci.appveyor.com/tools/encrypt) rather than including them directly in your repo.

One thing that would still be really nice is the feature described in issue #128. That would allow us to more easily do conditional deployment and take advantage of the cool stuff AppVeyor provides via deployment providers.

**Comment by mwhouser** [comment]
I'd like it as a deployment environment:

1. So that I don't have to have credentials (encrypted or not) in my repo, and
2. So that it can be done conditionally external to the build

**Comment by danielmcormond** [comment]
Yeah that would be sweet. We would use that too!

**Comment by keremdemirer** [comment]
We are deploying to elastic beanstalk succesfully but there is a problem that breaks continious integration with this method. After running the scripts you've mentioned, the response from aws is not includes if the application is deployed succesfully. It may take a while for the applicaiton be deployed to instances regarding to your configuration, or fail. So we cannot run acceptance tests for the deployments we make. 

There must be a better solution that confirms if the deployment is succesful. AWS SDK provides these informations, but I have no idea how to this with powershell.

It should be nice id appveyor provides feature this out of the box :)

**Comment by NuclearFishin** [comment]
Agree with @mwhouser that it would be nice to have EB deploys external to the build. 

For anyone looking for a workaround, I was able to achieve this by

* installing the AppVeyor deploy agent on a separate EC2 machine (with AWS tools installed)
* packaging the web deploy package inside an additional .zip build artifact
* deploying the .zip artifact as a "Windows Application", with a `deploy.ps1` script that invokes awsdeploy

**Comment by onuralp** [comment]
If there is a solution, that using elastic beanstalk deployment as an environment, we will use Appveyor.

**Comment by onuralp** [comment]
@NuclearFishin  Can you explain how can I invoke deploy.ps1 from seperated EC2 machine after deployment.

**Comment by NuclearFishin** [comment]
Sure, the `deploy.ps1` file is invoked automatically after you deploy- see the docs [here](http://www.appveyor.com/docs/deployment/agent#running-powershell-scripts-on-target-server-during-deployment). Be sure the `deploy.ps1` file is at the root of your .zip file when you deploy.

So from your AppVeyor build, you want to create a .zip file that contains at least:

* `deploy.ps1`
* `your-elastic-beanstalk-deploy-config.txt`
* `YourWebDeployPackage.zip`

Then, from within `deploy.ps1`, you invoke `awsdeploy.exe` using a command similar to the one referenced above in this thread. (`awsdeploy.exe` is installed by default on Windows EC2 machines.)

At a high level, the process goes a little something like this:

1. Build and package your source into a .zip file containing the files above
2. Configure a separate EC2 machine as an [AppVeyor deployment agent](http://www.appveyor.com/docs/deployment/agent)
3. Configure AppVeyor to deploy to this build agent as a [Windows application](http://www.appveyor.com/docs/deployment/agent#deploying-artifact-package-as-a-windows-application)

Then, when you deploy, AppVeyor will push the .zip file to the agent, unzip it, and run `deploy.ps1`, which will invoke `awsdeploy.exe` and deploy your web package.

Hope that helps, if you get really stuck, I can provide some sample scripts but please give me some time to clean them up for general use. (Might be best to do this as a wiki page so everyone can benefit)

**Comment by onuralp** [comment]
@NuclearFishin  I'll try this as soon as possible. It'll be nice complete article about this issue, if you have time.

**Comment by keremdemirer** [comment]
Lack of this feature breaks my ci flow. Hope to see a solution soon.

**Comment by FeodorFitsner** [comment]
Sample script for deployment:

```posh
$packageweb = $artifacts.values | Where-Object { $_.path -like '*App.Web.zip' }
$exe = "C:\Program Files (x86)\AWS Tools\Deployment Tool\awsdeploy.exe"

&$exe -r "-DDeploymentPackage=$($packageweb.path)" "-DTemplate=ElasticBeanstalk" "-DEnvironment.Name=MyAppWeb-test" "-DApplication.Name= MyAppWeb" "-DRegion=eu-west-1" "-DKeyPair=BlaKeypair" "-DUploadBucket=elasticbeanstalk-bucket" "-DAWSAccessKey=XXXXXXXXXXXXXXXXX" "-DAWSSecretKey=XXXXXXXXXXXXXXXXXXXXXX" "C:\projects\conductor\Conductor\Conductor.Web\awsdeploy.txt"
```

awsdeploy.txt :

```
Template = ElasticBeanstalk
Container.ApplicationHealthcheckPath = /healthcheck
```

**Comment by keremdemirer** [comment]
Thanks Feodor. This is the script I use currently. But suppports only one deployment which is not suitable for me. We need environment support for aws.

**Comment by FeodorFitsner** [comment]
Yeah, I've pasted it here for implementation.

I'm looking at the [guide here](http://docs.aws.amazon.com/AWSToolkitVS/latest/UserGuide/tkv-deployment-tool.html) - as far as I understand the deployment of BeanStalk app is basically just calling `awsdeploy.exe` tool with configuration file on input. Plus we could add like two options to the provider: `redeploy` and `update_stack`.

**Comment by keremdemirer** [comment]
A simple script deployment environment support should cover this I think. We can implement our custom script to call awsdeploy.exe.

**Comment by FeodorFitsner** [comment]
Look at this guide for building kind of "script" environment: http://www.appveyor.com/blog/2015/11/04/deployment-projects

**Comment by beardedprojamz** [comment]
Going off of [this comment](https://github.com/appveyor/ci/issues/45#issuecomment-165571187) - So, if I wanted to push to one ElasticBeanstalk environment for one branch (master -> dev), and another environment for another branch (live -> prod), is that something I'd need to do in PowerShell? Would I just rely on environment variables to figure out the branch I'm building?

**Comment by FeodorFitsner** [comment]
Correct.

**Comment by mycreativity** [comment]
+1

**Comment by IlyaFinkelshteyn** [comment]
Please check [this blog post](https://www.appveyor.com/blog/2016/11/07/aws-elastic-beanstalk/) about how to implement Elastic BeanStalk deployments with custom script.

**Comment by dustinmoris** [comment]
Hi it would be awesome to get this part of the deployment environments. We are a commercial AppVeyor customer who has a lot of projects that we need deploy to different Elastic Beanstalk environments. All the suggested hacks would force us to create a huge amount of projects. For example if we have 10 web services and 4 environments (dev, test, staging, production) then we end up with 40 projects to make the deployments possible instead of having only 10 projects with different deployment steps. This is a big issue for us. Is there any chance to get the Elastic Beanstalk deployment supported like Azure is done?

**Comment by keremdemirer** [comment]
@dustinmoris  Seems AppVeyor is not developing the product no more.  

I'd suggest you to move Visual Studio Team services. It do not provide AWS deployments out of the box but you can create custom tasks and re-use them with variables very easily. Plus you can run AppVeyor inside the VSTS which enables you to build your current configs. https://www.visualstudio.com/en-us/docs/marketplace/integrate/service-hooks/services/appveyor

**Comment by FeodorFitsner** [comment]
@dustinmoris as we noted earlier `awsdeploy.exe` tool and the library it's based on are not secure to use in multi-tenant shared environment which current "Environments" is. However, year 2016 gave us Docker for Windows Containers and we are looking into using it for building new lightweight and isolated "Environments" v2 that would allow customers to run ElasticBeanstalk tools and/or any custom PowerShell scripts to do their deployments.

For now the recommended way is doing ElasticBeanstalk deployments in a separate projects/builds which guarantees isolation. I'm not sure I understand why the number of deployment projects is a concern here as you'd have the same number of "environments". You can actually have few "deployment" projects and then customize them with environment variables either in "run-time" (passing them when starting a new build of deployment project via API) or "design-time" (specifying variables on project settings UI - they will be added/overridden to those ones defined in appveyor.yml).

@keremdemirer Not having a provider for ElasticBeanstalk doesn't mean we are not developing the product. AppVeyor is a complex *service* with a lot of moving parts. We've seen its rapid growth recently and we were addressing various scaling challenges. NuGet feeds hosting is one of them - we are rolling out a new implementation which is in some cases 100-1000 times faster then previous one. Another cool feature we were working on is [private build clouds](https://www.appveyor.com/docs/enterprise/running-builds-on-azure/) which is not widely advertised yet. To allow AppVeyor customers running builds on their own Azure, GCE, AWS, Hyper-V and Docker clouds nearly 60% of backend code was refactored!

Good luck with VSTS - be aware though they don't run builds on dedicated VMs with admin rights, so your abilities to install additional software or configure the system are limited.

**Comment by dustinmoris** [comment]
@FeodorFitsner Cool thanks for your answer. The approach of few deployment projects with customizable settings through environment variables is something that could work for us (need to think more about how it fits into our current work flow though). I also like the idea of a lighter version of environments with Docker containers!

I had another idea today, which I thought would be a nice feature as well and would love to hear your opinion on it:

What if I could set up multiple "build_script" or "deployment" steps in the appveyor.yml file. One would always be the default and run on a commit, but the others would be something that could be manually kicked off from the UI or API after the default build has succeeded. This way someone could have multiple additional script steps in one appveyor.yml file to do additional things like start a deployment or promote a deployment, kick off some long running automation tests that you wouldn't want to run every time, or other useful stuff.

A feature like this would make environments and deployments almost redundant and it could run the same way that a normal build already runs, isolated on the same or a new build worker. I guess this would be somehow more flexible than Environments v2 and at the same time a lot less work I guess?

**Comment by IlyaFinkelshteyn** [comment]
@dustinmoris If I understand your proposal correctly, you want deploy script to behave like one of providers for you to be able to add multiple of them and make then run conditionally, right?

But why you don't like to have single deployment script which behaves differently because of condition, like:

```
if ($env:APPVEYOR_FORCED_BUILD) # started from API or UI
{
    # run long running tests
}
if($env:CONFIGURATION -eq "Release")
{
    # deploy to production
}
```

I believe that any deployment condition can be expressed with [AppVeyor environment variables](https://www.appveyor.com/docs/environment-variables/). If you see some condition which you cannot express this way, please let us know.

If you don't like to have this baked in the script, but have logic readable from YAML, you can do something like

```
deploy_script:
- ps: if ($env:APPVEYOR_FORCED_BUILD) {.\RunLongRunningTests.ps1}
- ps: if ($env:CONFIGURATION -eq "Release") {.\DeployToProductionps1}
```

Let us know what you think.

@keremdemirer I just can't stop myself and add couple of links to what Feodor said:
https://www.appveyor.com/updates/
https://www.appveyor.com/blog/2016/09/28/the-new-build-cache/

**Comment by dustinmoris** [comment]
@IlyaFinkelshteyn Could you elaborate how I would set environment variables before starting a new build then? Is this something I would have to change in the UI settings page?

If I have to change it in the Settings page, then I assume this setting would persist until I remove it again? This would be a big problem, because it would sort of defeat the purpose of a CI. A good CI tool should make it easy to reduce any potential human error. If I change lets say an env var to `DEPLOY_TARGET=Production` and then forget to change it back to `DEPLOY_TARGET=Development` then the next person who kicks off the build might accidentally deploy something to production which was never meant to be.

Also how can I set up conditional notifications with that strategy? On normal builds I only want a notification to occur when the build status changes, but on deployments to production or staging I want the build to always notify a certain Slack channel.

Also I don't understand how with any of those strategies I can control which build artifacts should be deployed. Let's say I have 3 commits which triggered 3 different builds simultaneously. Now I want to deploy the oldest of the 3 builds to dev, then promote to staging and then promote to production. How can I easily specify which build artifacts should be used for the deployment?

If there is a good way to do all of this then please let me know and I'll definitely give it a try.

**Comment by FeodorFitsner** [comment]
The idea is to have several ["deployment" projects](https://www.appveyor.com/blog/2015/11/04/deployment-projects/) based on the same repository. Say, for deployment "environment" you created `myapp-deployment` repository. Somewhere in that repository there is a script doing the deployment that rely on `server_name` environment variable.

Then you add two new projects in AppVeyor each time selecting the same `myapp-deployment` repo. Rename first project to "Deploy myapp to Development" and second to "Deploy myapp to Production". Then on `Environment` tab of first project settings configure `server_name` environment variable with the name of development server and for "Deploy myapp to Production" project set `server_name` to the name of production server.

As you are authoring a custom script for deployment you can control which artifacts are getting deployed with AppVeyor API.

I understand all this might be quite overwhelming, but we are looking into improving AppVeyor's deployment story. Though it would never cover all the cases. You may consider using some specialized deployment tools together with AppVeyor like Octopus Deploy.

**Comment by IlyaFinkelshteyn** [comment]
Elastic Beanstalk deployment 1st class support is now added. Please check [documentation](https://www.appveyor.com/docs/deployment/amazon-elastic-beanstalk/).

**Comment by darryljjennings** [comment]
Thanks @dustinmoris !
